### PR TITLE
Add service proc macro

### PR DESCRIPTION
**Pull request by petekubiak**
The purpose of this PR is to simplify the process for defining services and characteristics in TrouBLE, as laid out in this issue: #130.
Services are defined as a struct decorated with the `#[gatt_service(uuid = ...)]` attribute and fields within the struct can be decorated with `#[characteristic(<args>)]` attributes to define them as characteristics.
In a similar way, the gatt server is defined as a struct with the `#[gatt_server]` attribute, where each of the fields is a service.
The server is then instantiated with the `::new()` function which is generated by the attribute macro.

This PR lays the foundation for expanding the functionality of the server, services and characteristics such that all access to characteristics (e.g. set, get, notify, indicate) can be done through methods on the characteristics, accessed through the server.

### CODE REVIEW

**Comment by plaes** [comment]
@petekubiak Could you split out the uuid parser/generator macro as separate PR?

This would immensely improve implementing services with 128-bit Uuids:
```rust
    let mut uart_service = table
        // #[gatt_service(uuid = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E")] 
        .add_service(Service::new(Uuid::Uuid128([
                    0x9e, 0xca, 0xdc, 0x24,
                    0x0e, 0xe5, 0xa9, 0xe0,
                    0x93, 0xf3, 0xa3, 0xb5,
                    0x01, 0x00, 0x40, 0x6e,
            ])));

    // #[characteristic(uuid = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E", write)]
    let _ = uart_service.add_characteristic(Uuid::Uuid128([
                    0x9e, 0xca, 0xdc, 0x24,
                    0x0e, 0xe5, 0xa9, 0xe0,
                    0x93, 0xf3, 0xa3, 0xb5,
                    0x02, 0x00, 0x40, 0x6e,
            ]),
            &[
                CharacteristicProp::Write,
            ],
            &mut rx,
        );
```

**Comment by alexmoon** [comment]
You can write `.add_service(Service::new(Uuid::Uuid128(0x6E400001_B5A3_F393_E0A9_E50E24DCCA9Eu128.to_le_bytes())))`. No macro needed.

**Comment by petekubiak** [comment]
Following on from that my hope is that you wouldn't even need to call `.add_service()` or `.add_characteristic()` as the attribute macro takes care of that for you when you call `<Server name>::new()` (Server name being the name of the struct you have decorated with the `#[gatt_server]` attribute).

File: host-macro/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "trouble-host-macro"
```
Comment (lulf) [inline host-macro/Cargo.toml:2]: Maybe name this 'trouble-host-macros', as I think it can be used for more than the derive macros.

**Review by lulf — COMMENTED** [review]
Looks great, thanks the work on this! I've left a few inline comments.

I noticed a commend saying the pub server field would be non-public with additional changes. Do you wish to merge this now or make additional changes to it?

**Comment by petekubiak** [comment]
> Looks great, thanks the work on this! I've left a few inline comments.
> 
> I noticed a commend saying the pub server field would be non-public with additional changes. Do you wish to merge this now or make additional changes to it?

I think we'd like to merge this one first and then do further development in a new PR (avoids this one getting too large :) )

File: Cargo.toml.hidden [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    "host",
+    "host-macro", 
+    "examples/apps", 
+    # "examples/esp32", 
+    # "examples/nrf-sdc",
+    # "examples/rp-pico-w",
+    # "examples/serial-hci",
+    # "examples/apache-nimble",
+]
+resolver = '2'
```
[the remark this answers is not in the document]
Comment (jamessizeland) [reply Cargo.toml.hidden:12]: yes, we can remove it.  Having the top level workspace helps with kicking the rust-analyser into action, but makes running the examples more awkward I think.  I'll remove it.

File: examples/apps/src/ble_bas_peripheral.rs [hunk 3]
```diff
@@ -90,8 +89,7 @@ async fn gatt_task<C: Controller>(
 
 async fn advertise_task<C: Controller>(
     mut peripheral: Peripheral<'_, C>,
-    server: &GattServer<'_, '_, C, NoopRawMutex, MAX_ATTRIBUTES, L2CAP_MTU>,
-    handle: Characteristic,
+    server: &Server<'_, '_, C, NoopRawMutex, MAX_ATTRIBUTES, L2CAP_MTU>,
```
[the remark this answers is not in the document]
Comment (jamessizeland) [reply examples/apps/src/ble_bas_peripheral.rs:92]: I like the sound of that.

[the remark this answers is not in the document]
Comment (jamessizeland) [reply examples/apps/src/ble_bas_peripheral.rs:92]: I suppose in the proposed future interface:
- our constants MAX_ATTRIBUTES and L2CAP_MTU can be inferred from the macro (I'm not particularly familiar with L2CAP yet...) if we're not needing to prebuild an AttributeTable.  
- NoopRawMutex could be the default and changeable as an arg on the gatt_server macro.
- Controller would need to stay generic I think?
- I'm not sure about the lifetimes, I think they get simplified too, but mught need some of the future PR #136 work first.

**Comment by petekubiak** [comment]
> Looks great, thanks the work on this! I've left a few inline comments.
> 
> I noticed a commend saying the pub server field would be non-public with additional changes. Do you wish to merge this now or make additional changes to it?

I've decided to set the visibilities back to what they were and instead implement getter functions for the fields I need. At least that way they are immutable.

File: examples/apps/src/ble_bas_peripheral.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (petekubiak) [reply examples/apps/src/ble_bas_peripheral.rs]: Implemented in 0f9801e. We've kept the controller as a generic as it sometimes needs to be inferred (e.g. in ble_bas_peripheral.rs) while sometimes it needs to be defined explicitly (e.g. the tests)

File: Cargo.toml.hidden (hunk 2 above)
[the remark this answers is not in the document]
Comment (petekubiak) [reply Cargo.toml.hidden]: Removed in 2c7f5d5983df0bfe0c7df10d264364709ee3cc28

File: examples/apps/src/ble_derive_service.rs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    #[characteristic(uuid = "2A39", write)]
+    control: u8,
+    #[characteristic(uuid = "2A63", read, notify)]
+    energy_expended: u16,
+}
+
+fn main() {
+    let mut table: AttributeTable<NoopRawMutex, 10> = AttributeTable::new();
+    let _service = HeartRateService::new(&mut table);
+}
```
[the remark this answers is not in the document]
Comment (petekubiak) [reply examples/apps/src/ble_derive_service.rs]: Removed in 106903ce96902bd801df26f5f4a5ea26ce89fcb5.
We will add an example demonstrating all features in the next PR.

File: host-macro/Cargo.toml (hunk 1 above)
Comment (petekubiak) [reply host-macro/Cargo.toml]: Renamed in 87b7dae6a8ea9bd02212610476f57c731c7b179b

**Review by lulf — APPROVED** [review]
Looks good to me!

**Comment by lulf** [comment]
The new integration test seems to be failing, does it pass locally? https://github.com/embassy-rs/trouble/actions/runs/11340899524/job/31538139912

**Comment by petekubiak** [comment]
> The new integration test seems to be failing, does it pass locally? https://github.com/embassy-rs/trouble/actions/runs/11340899524/job/31538139912

Aha, we don't actually have a rig to run locally (I think I'll try to get one set up). We thought that the tests would run automatically once the PR was converted from a draft.

**Comment by petekubiak** [comment]
Thanks for the information, and thank you for adding me!

**Comment by petekubiak** [comment]
Would it be possible to add @jamessizeland as well?

**Comment by lulf** [comment]
Looks like I broke the bot config, I need to look into that. For now I've triggered it manually.

**Comment by jamessizeland** [comment]
ok still an issue to fix looks like!

**Review by jamessizeland — APPROVED** [review]

**Pull request merged**